### PR TITLE
Add New NavView VisualStateGroup 

### PR DESCRIPTION
**Pull request by karenbtlai**
## Description

- Add `PaneOverlayGroup` to control `PaneOverlay` and `PaneNotOverlay` vistual states
- Set background to transparent when in compact closed, and acrylic when in compact overlay
- Lift `SplitView` from WUXC to add cornerRadius templateBindings to `PaneRoot`
- Set `OverlayCornerRadius` to  SplitView during `PaneOverlay` state
- Update NavView pane border to have rounded corners on overlay

## How Has This Been Tested?

Visual verification

## Screenshots (if appropriate):

image

### CODE REVIEW

File: dev/CommonStyles/Common_themeresources.xaml [hunk 1]
```diff
@@ -56,10 +56,6 @@
 
     <x:Boolean x:Key="ThemeShadowIsUsingDropShadows">True</x:Boolean>
 
-    
-    <controls:TextCommandBarFlyout x:Key="TextControlCommandBarContextFlyout" Placement="BottomEdgeAlignedLeft" />
-    <controls:TextCommandBarFlyout x:Key="TextControlCommandBarSelectionFlyout" Placement="TopEdgeAlignedLeft" />
```
[the remark this answers is not in the document]
Comment (karenbtlai) [reply dev/CommonStyles/Common_themeresources.xaml:61]: oops, this was temporarily removed since it was causing build failures in InnerLoop. I guess I accidentally checked it in, I will revert this change.

File: dev/NavigationView/NavigationView.xaml [hunk 2]
```diff
@@ -18,36 +18,9 @@
             <Setter.Value>
                 <ControlTemplate TargetType="local:NavigationView">
                     <Grid x:Name="RootGrid">
-                        <Grid.Resources>
-                            <Storyboard x:Name="ShadowCasterEaseInStoryboard">
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationView.xaml:22]: @ranjeshj Is this safe to delete? technically it is a named resource..

Comment (StephenLPeters) [reply dev/NavigationView/NavigationView.xaml:22]: @MikeHillberg FYI

---
In reply to: [605142979](https://github.com/microsoft/microsoft-ui-xaml/pull/4659#discussion_r605142979) [](ancestors = 605142979)

File: dev/NavigationView/NavigationView.xaml [hunk 3]
```diff
@@ -118,6 +91,53 @@
...
+                                            <DoubleAnimationUsingKeyFrames Storyboard.TargetName="ShadowCaster" Storyboard.TargetProperty="Opacity">
+                                                <LinearDoubleKeyFrame KeyTime="0:0:0.12" Value="0" />
+                                            </DoubleAnimationUsingKeyFrames>
+                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ShadowCaster" Storyboard.TargetProperty="HorizontalAlignment">
+                                                <DiscreteObjectKeyFrame KeyTime="0" Value="Left"/>
+                                            </ObjectAnimationUsingKeyFrames>
+                                        </Storyboard>
+                                    </VisualTransition>
+                                </VisualStateGroup.Transitions>
+                                <VisualState x:Name="PaneOverlay">
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationView.xaml:125]: These VisualState names don't seem correct to me. @ranjeshj and @MikeHillberg for naming, maybe PaneOverlaying and PaneNotOverlaying

File: dev/NavigationView/NavigationView.xaml [hunk 4]
```diff
@@ -118,6 +91,53 @@
...
+                                    <VisualState.Setters>
+                                        <Setter Target="RootSplitView.CornerRadius" Value="{ThemeResource OverlayCornerRadius}" />
+                                        <Setter Target="RootSplitView.BorderThickness" Value="0,0,1,0" />
+                                        <Setter Target="ShadowCaster.Opacity" Value="1" />
+                                    </VisualState.Setters>
+                                </VisualState>
+                                <VisualState x:Name="PaneNotOverlay">
+                                    <VisualState.Setters>
+                                        <Setter Target="RootSplitView.PaneBackground" Value="{ThemeResource NavigationViewExpandedPaneBackground}" />
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationView.xaml:134]: One of these VisualStates should be the default state and should not have any setters in it. Instead these properties should be set on the element directly.

File: dev/NavigationView/NavigationView.xaml [hunk 5]
```diff
@@ -428,10 +450,11 @@
                                 <SplitView.Pane>
                                     <Grid 
                                         x:Name="PaneContentGrid"
+                                        BorderBrush="{ThemeResource NavigationViewItemSeparatorForeground}"
+                                        BorderThickness="0"
                                         HorizontalAlignment="Left"
                                         Visibility="{Binding RelativeSource={RelativeSource TemplatedParent}, Path=TemplateSettings.LeftPaneVisibility}"
-                                        BorderBrush="{ThemeResource NavigationViewItemSeparatorForeground}"
-                                        BorderThickness="0,0,1,0">
+                                        contract7Present:CornerRadius="0">
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationView.xaml:457]: Why is this needed?

File: dev/NavigationView/NavigationView.cpp [hunk 6]
```diff
@@ -1714,6 +1719,14 @@ void NavigationView::OnSplitViewPaneOpening(const winrt::DependencyObject& /*sen
     {
         // See UpdateIsClosedCompact 'RS3+ animation timing enhancement' for explanation:
         winrt::VisualStateManager::GoToState(*this, L"ListSizeFull", true /*useTransitions*/);
+
+        if (auto splitView = m_rootSplitView.get())
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationView.cpp:1723]: nit: const

File: dev/NavigationView/NavigationView.xaml (hunk 2 above)
Comment (karenbtlai) [reply dev/NavigationView/NavigationView.xaml:22]: That was the storyboard I added for the NavigationView Shadow PR, I don't know if that was checked in for an actual release or preview though. However, I moved it into `PaneOverlayGroup`'s VisualTransition bit and kept the name just in case. Should that be removed?

Comment (StephenLPeters) [reply dev/NavigationView/NavigationView.xaml:22]: ah, okay I think that's okay then

File: dev/NavigationView/NavigationView.xaml (hunk 5 above)
Comment (karenbtlai) [reply dev/NavigationView/NavigationView.xaml]: oops, reverted.

**Comment by ranjeshj** [comment]
@karenbtlai you need to merge in master to fix the key mismatch errors.

File: dev/NavigationView/NavigationView.cpp [hunk 7]
```diff
@@ -1698,6 +1698,11 @@ void NavigationView::OnSplitViewPaneClosing(const winrt::DependencyObject& /*sen
                     winrt::VisualStateManager::GoToState(*this, L"ListSizeCompact", true /*useTransitions*/);
                     UpdatePaneToggleSize();
                 }
+
+                if (splitView.DisplayMode() == winrt::SplitViewDisplayMode::CompactOverlay || splitView.DisplayMode() == winrt::SplitViewDisplayMode::Overlay)
+                {
+                    winrt::VisualStateManager::GoToState(*this, L"PaneNotOverlay", true /*useTransitions*/);
+                }
```
Comment (licanhua) [inline dev/NavigationView/NavigationView.cpp:1705]: It's very comfusion here.
From your code, Likely PaneNotOverlay and PaneOverlay only happens when DisplayMode is CompactOverlay or Overlay.
So we should have 3 visual states other than 2:
1. DisplayMode is not CompactOverlay or Overlay
2. DisplayMode is ?Overlay, and Pane is closed
3. DisplayMode is ?Overlay, and Pane is opened

Don't use PaneNotOverlay/Overlay, it's easy to mess up with the term: CompactOverlay or Overlay in splitview setting.

It's better to have the LeftNav in the name, for example: Pane???OnLeftNavigationView

**Comment by StephenLPeters** [comment]
> static constexpr auto c_shadowCasterEaseOutStoryboard = L"ShadowCasterEaseOutStoryboard"sv;

Looks like these storyboards aren't referenced from the code behind anymore. Can we delete this? If so, can we remove the name from the storyboard definition as well? (We can only do that if we haven't published the storyboard in a release yet)

---
Refers to: dev/NavigationView/NavigationView.cpp:101 in 3e2d182. [](commit_id = 3e2d182425a426107bd8fd6d22feab0dc2e2ff09, deletion_comment = False)

**Comment by StephenLPeters** [comment]
>                 UnsetDropShadow();

Don't we need to keep the shadow around until the transition animation has finished?

---
Refers to: dev/NavigationView/NavigationView.cpp:4086 in 3e2d182. [](commit_id = 3e2d182425a426107bd8fd6d22feab0dc2e2ff09, deletion_comment = False)

File: dev/NavigationView/NavigationView.xaml [hunk 8]
```diff
@@ -595,7 +614,7 @@
                                 x:Name="ShadowCaster"
                                 Grid.RowSpan="2"
                                 Width="{TemplateBinding OpenPaneLength}"
-                                Opacity="0"
+                                Opacity="1"
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationView.xaml:617]: >Opacity [](start = 32, length = 7)

Opacity of 1 is the default, can we remove this line?

**Comment by karenbtlai** [comment]
> > ```
> >             UnsetDropShadow();
> > ```
> 
> Don't we need to keep the shadow around until the transition animation has finished?
> 
> Refers to: dev/NavigationView/NavigationView.cpp:4086 in [3e2d182](https://github.com/microsoft/microsoft-ui-xaml/commit/3e2d182425a426107bd8fd6d22feab0dc2e2ff09). [](commit_id = [3e2d182](https://github.com/microsoft/microsoft-ui-xaml/commit/3e2d182425a426107bd8fd6d22feab0dc2e2ff09), deletion_comment = False)

You're right. I reverted the bit with the eventlistener

File: dev/NavigationView/NavigationView.cpp [hunk 9]
```diff
@@ -686,9 +680,6 @@ void NavigationView::OnApplyTemplate()
     if (SharedHelpers::Is21H1OrHigher())
     {
         m_shadowCaster.set(GetTemplateChildT<winrt::Grid>(c_shadowCaster, controlProtected));
-        m_shadowCasterEaseInStoryboard.set(GetTemplateChildT<winrt::Storyboard>(c_shadowCasterEaseInStoryboard, controlProtected));
```
Comment (StephenLPeters) [inline dev/NavigationView/NavigationView.cpp:689]: We also need to delete these from the header file right?

Comment (karenbtlai) [reply dev/NavigationView/NavigationView.cpp:689]: removed!

**Review by StephenLPeters — APPROVED** [review]
:shipit:

File: dev/NavigationView/NavigationView.cpp (hunk 7 above)
Comment (karenbtlai) [reply dev/NavigationView/NavigationView.cpp:1700]: Added you to the email thread discussing the naming

File: dev/NavigationView/NavigationView.xaml [hunk 10]
```diff
@@ -118,6 +91,49 @@
...
+                                            <DoubleAnimationUsingKeyFrames Storyboard.TargetName="ShadowCaster" Storyboard.TargetProperty="Opacity">
+                                                <LinearDoubleKeyFrame KeyTime="0:0:0.12" Value="0" />
+                                            </DoubleAnimationUsingKeyFrames>
+                                            <ObjectAnimationUsingKeyFrames Storyboard.TargetName="ShadowCaster" Storyboard.TargetProperty="HorizontalAlignment">
+                                                <DiscreteObjectKeyFrame KeyTime="0" Value="Left"/>
+                                            </ObjectAnimationUsingKeyFrames>
+                                        </Storyboard>
+                                    </VisualTransition>
+                                </VisualStateGroup.Transitions>
+                                <VisualState x:Name="PaneOverlaying" />
```
Comment (licanhua) [inline dev/NavigationView/NavigationView.xaml:125]: Please do a testing that if the default NavigationVIew is set to minimal or compact in Xaml, will PaneNotOverlaying 
 code still called?
Because your default state is PaneOverlaying, and I want to make sure PaneNotOverlaying working in some default settings.

Comment (karenbtlai) [reply dev/NavigationView/NavigationView.xaml:125]: Yes it does. I confirmed it with different test pages where minimal and compact is the default.
